pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/ImmutableFST.java

```diff
@@ -0,0 +1,406 @@
+    return decodeFromBytes(node, _nodeDataLength);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>
+   * For this automaton version, an additional {@link FSTFlags#NUMBERS} flag may
+   * be set to indicate the automaton contains extra fields for each node.
+   * </p>
+   */
+  @Override
+  public Set<FSTFlags> getFlags() {
+    return _flags;
+  }
+
+  /**
+   * Returns <code>true</code> if this arc has <code>NEXT</code> bit set.
+   *
+   * @see #BIT_LAST_ARC
+   * @param arc The node's arc identifier.
+   * @return Returns true if the argument is the last arc of a node.
+   */
+  public boolean isArcLast(int arc) {
+    return (getByte(arc, ADDRESS_OFFSET) & BIT_LAST_ARC) != 0;
+  }
+
+  /**
+   * @see #BIT_TARGET_NEXT
+   * @param arc The node's arc identifier.
+   * @return Returns true if {@link #BIT_TARGET_NEXT} is set for this arc.
+   */
+  public boolean isNextSet(int arc) {
+
+    return (getByte(arc, ADDRESS_OFFSET) & BIT_TARGET_NEXT) != 0;
+  }
+
+  /**
+   * Returns an n-byte integer encoded in byte-packed representation.
+   */
+  final int decodeFromBytes(final int start, final int n) {
+    int r = 0;
+
+    for (int i = n; --i >= 0; ) {
+      Pair<Integer, Integer> offheapOffsets = getOffheapOffsets(start + i);
+      byte[] inputData = _mutableBytesStore.get(offheapOffsets.getFirst());
+
+      r = r << 8 | (inputData[offheapOffsets.getSecond()] & 0xff);
+    }
+    return r;
+  }
+
+  /**
+   * Returns the address of the node pointed to by this arc.
+   */
+  final int getDestinationNodeOffset(int arc) {
+    if (isNextSet(arc)) {
+      /* The destination node follows this arc in the array. */
+      return skipArc(arc);
+    } else {
+      /*
+       * The destination node address has to be extracted from the arc's
+       * goto field.
+       */
+      return decodeFromBytes(arc + ADDRESS_OFFSET, _gotoLength) >>> 3;
+    }
+  }
+
+  /**
+   * Read the arc's layout and skip as many bytes, as needed.
+   */
+  private int skipArc(int offset) {
+    return offset + (isNextSet(offset) ? 1 + 1   /* label + flags */ : 1 + _gotoLength /* label + flags/address */);
+  }
+
+  private byte getByte(int seek, int offset) {
+    Pair<Integer, Integer> offheapOffsets = getOffheapOffsets(seek);
+
+    int fooArc = offheapOffsets.getFirst();
+    byte[] retVal = _mutableBytesStore.get((fooArc));
+
+    int barArc = offheapOffsets.getSecond();
+    int target = barArc + offset;
+
+    if (target >= PER_BUFFER_SIZE) {
+      retVal = _mutableBytesStore.get(fooArc + 1);
+      target = target - PER_BUFFER_SIZE;
+    }
+
+    return retVal[target];
+  }
+
+  private Pair<Integer, Integer> getOffheapOffsets(int seek) {
+    int fooArc = seek >= PER_BUFFER_SIZE ? seek / PER_BUFFER_SIZE : 0;
+    int barArc = seek >= PER_BUFFER_SIZE ? seek - ((fooArc) * PER_BUFFER_SIZE) : seek;
+
+    assert fooArc < _mutableBytesStore.getNumValues();
+    assert barArc < PER_BUFFER_SIZE;
+
+    return new Pair<>(fooArc, barArc);
+  }
```

Comment:
I would be tempted to inline this, it's only used twice (but measure the impact of doing so)

Reply:
Since barArc depends on fooArc, is inlining possible here?

Reply:
What I mean is copy and paste the lines 
```java
    int fooArc = seek >= PER_BUFFER_SIZE ? seek / PER_BUFFER_SIZE : 0;
    int barArc = seek >= PER_BUFFER_SIZE ? seek - ((fooArc) * PER_BUFFER_SIZE) : seek;
```

to where you need them (2 places) instead of factoring this in to a method, at the cost of boxing `fooArc` and `barArc` and wrapping them in a `Pair` so they can be returned.

So, for example, in `decodeFromBytes` you would write

```java
  final int decodeFromBytes(final int start, final int n) {
     int r = 0;

     for (int i = n; --i >= 0; ) {
       int seek = start + i;
       int fooArc = seek >= PER_BUFFER_SIZE ? seek / PER_BUFFER_SIZE : 0;
       int barArc = seek >= PER_BUFFER_SIZE ? seek - ((fooArc) * PER_BUFFER_SIZE) : seek;

       byte[] inputData = _mutableBytesStore.get(fooArc);

       r = r << 8 | (inputData[barArc] & 0xff);
     }
     return r;
   }
```

And similar in the other place this method is called.

Reply:
Got it, thanks for the explanation. I was under the impression that we were trying to avoid storing the value of the offsets in actual integers (to avoid boxing). Fixed, thanks